iroh-net/src/magicsock/node_map/udp_paths.rs

```diff
@@ -0,0 +1,148 @@
+    chosen_candidate: Option<IpPort>,
+}
+
+impl NodeUdpPaths {
+    pub(super) fn new() -> Self {
+        Default::default()
+    }
+
+    #[cfg(test)]
+    pub(super) fn from_parts(paths: BTreeMap<IpPort, PathState>, best_addr: BestAddr) -> Self {
+        Self {
+            paths,
+            best_addr,
+            chosen_candidate: None,
+        }
+    }
+
+    /// Returns the current UDP address to send on.
+    ///
+    /// TODO: The goal here is for this to simply return the already known send address, so
+    /// it should be `&self` and not `&mut self`.  This is only possible once the state from
+    /// [`NodeUdpPaths`] is no longer modified from outside.
+    pub(super) fn send_addr(&mut self, now: Instant, have_ipv6: bool) -> UdpSendAddr {
+        self.assign_best_addr_from_candidates_if_empty();
+        match self.best_addr.state(now) {
+            best_addr::State::Valid(addr) => UdpSendAddr::Valid(addr.addr),
+            best_addr::State::Outdated(addr) => UdpSendAddr::Outdated(addr.addr),
+            best_addr::State::Empty => {
+                // No direct connection has been used before.  If we know of any possible
+                // candidate addresses, randomlly try to use one.  This path is most
+                // effective when folks use a NodeAddr with exactly one direct address which
+                // they know to work, effectively like using a traditional socket or quic
+                // endpoint.
+                let addr = self
+                    .chosen_candidate
+                    .and_then(|ipp| self.paths.get(&ipp))
+                    .and_then(|path| path.udp_addr())
+                    .filter(|addr| addr.is_ipv4() || have_ipv6)
+                    .or_else(|| {
+                        // Look for a new candidate in all the known paths.  This may look
+                        // like a RNG use on the hot-path but this is normally invoked at
+                        // most most once at startup.
+                        let addr = self
+                            .paths
+                            .values()
+                            .filter_map(|path| path.udp_addr())
+                            .filter(|addr| addr.is_ipv4() || have_ipv6)
+                            .choose(&mut rand::thread_rng());
+                        self.chosen_candidate = addr.as_ref().map(|addr| IpPort::from(*addr));
+                        addr
+                    });
+                match addr {
+                    Some(addr) => UdpSendAddr::Unconfirmed(addr),
+                    None => UdpSendAddr::None,
+                }
+            }
+        }
+    }
+
+    /// Fixup best_adrr from candidates.
+    ///
+    /// If somehow we end up in a state where we failed to set a best_addr, while we do have
+    /// valid candidates, this will chose a candidate and set best_addr again.  Most likely
+    /// this is a bug elsewhere though.
```

Comment:
if this is a bug, shouldn't we log that we are doing this, to figure out why this happens?

Reply:
It is already logged on `warn` for this purpose.